central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,166 @@
+)
+
+type datastoreImpl struct {
+	storage        postgres.Store
+	indicatorStore processIndicatorStore.DataStore
+}
+
+var (
+	plopSAC = sac.ForResource(resources.ProcessListeningOnPort)
+	log     = logging.LoggerForModule()
+)
+
+func newDatastoreImpl(
+	storage postgres.Store,
+	indicatorDataStore processIndicatorStore.DataStore,
+) *datastoreImpl {
+	return &datastoreImpl{
+		storage:        storage,
+		indicatorStore: indicatorDataStore,
+	}
+}
+
+func (ds *datastoreImpl) AddProcessListeningOnPort(
+	ctx context.Context,
+	portProcesses ...*storage.ProcessListeningOnPort,
+) error {
+	defer metrics.SetDatastoreFunctionDuration(
+		time.Now(),
+		"ProcessListeningOnPort",
+		"AddProcessListeningOnPort",
+	)
+
+	var (
+		lookups       []*v1.Query
+		indicatorsMap = map[string]*storage.ProcessIndicator{}
+	)
+
+	if ok, err := plopSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	// First query all needed process indicators in one go
+	for _, val := range portProcesses {
+		lookups = append(lookups,
+			search.NewQueryBuilder().
+				AddExactMatches(search.ContainerName, val.Process.ContainerName).
+				AddExactMatches(search.PodID, val.Process.PodId).
+				AddExactMatches(search.ProcessName, val.Process.ProcessName).
+				AddExactMatches(search.ProcessArguments, val.Process.ProcessArgs).
+				AddExactMatches(search.ProcessExecPath, val.Process.ProcessExecFilePath).
+				ProtoQuery())
+	}
+
+	indicatorsQuery := search.DisjunctionQuery(lookups...)
+	log.Debugf("Sending query: %s", indicatorsQuery.String())
+	indicators, err := ds.indicatorStore.SearchRawProcessIndicators(ctx, indicatorsQuery)
+	if err != nil {
+		return err
+	}
+
+	for _, val := range indicators {
+		key := fmt.Sprintf("%s %s %s %s %s",
+			val.GetContainerName(),
+			val.GetPodId(),
+			val.GetSignal().GetName(),
+			val.GetSignal().GetArgs(),
+			val.GetSignal().GetExecFilePath(),
+		)
+
+		// A bit of paranoia is always good
+		if old, ok := indicatorsMap[key]; ok {
+			log.Warnf("An indicator %s is already present, overwrite with %s",
+				old.GetId(), val.GetId())
+		}
+
+		indicatorsMap[key] = val
+	}
+
+	plopObjects := make([]*storage.ProcessListeningOnPortStorage, len(portProcesses))
+	for i, val := range portProcesses {
+		indicatorID := ""
+
+		key := fmt.Sprintf("%s %s %s %s %s",
+			val.Process.ContainerName,
+			val.Process.PodId,
+			val.Process.ProcessName,
+			val.Process.ProcessArgs,
+			val.Process.ProcessExecFilePath,
+		)
+
+		if indicator, ok := indicatorsMap[key]; ok {
+			indicatorID = indicator.GetId()
+			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
+		} else {
+			// XXX: Create a metric for this
+			log.Warnf("Found no matching indicators for %s", key)
+		}
+
+		plopObjects[i] = &storage.ProcessListeningOnPortStorage{
+			// XXX, ResignatingFacepalm: Use regular GENERATE ALWAYS AS
+			// IDENTITY, which would require changes in store generator
```

Comment:
if we are going to keep the store custom, no need to restrict ourselves to this ID pattern if another pattern makes sense.  IMO

Reply:
It was harder to do without UUID support, when the feature was started. Now should be easier, but to limit the scope a bit I would add this as a follow up step (already got quite a list).

Reply:
the only point of caution on that, is that if we are going to change the ID we should do so before it gets deployed for people to use so we don't have to write a migration.  so follow up is fine as long as that happens before it gets deployed, IMO

Reply:
This change ideally would require only a schema migration, not the data migration, so maybe this will make things easier. The plan I have is to change the pk column and add a default value to generate random UUID.

Reply:
Sadly, changing the type of the PK would not be seamlessly handled by Gorm.  So there would be some work to do there.